[css-anchor-position-1] Implement swapping due to a try-tactic for position-area

tuankiet65 opened the pull request:
#### 0942066d2a6f912865487ac2d673bebd5178c13d
<pre>
[css-anchor-position-1] Implement swapping due to a try-tactic for position-area
<a href="https://rdar.apple.com/146814606">rdar://146814606</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=289589">https://bugs.webkit.org/show_bug.cgi?id=289589</a>

Reviewed by Antti Koivisto.

This commit implements swapping due to a try-tactic [1] for position-area.
This is done at style building time; BuilderConverter::convertPositionArea
will flip position-area as specified by the position-try fallback. This is
needed instead of flipping position-area at layout time, because
position-area should compute to the flipped value.

There are some remaining test failures in try-tactic-position-area.html;
these has to do with how position-area should be computed. The test expects
the computed value of position-area to use axis-ambiguous keywords, when the
specified value also uses those keywords (e.g: &apos;start end&apos; computes to
&apos;start end&apos;). However, WebKit currently computes to their axis-unambiguous
equivalent (e.g: &apos;start end&apos; computes to &apos;block-start block-end&apos;).
This issue is tracked in [2].

[1]: <a href="https://drafts.csswg.org/css-anchor-position-1/#swap-due-to-a-try-tactic">https://drafts.csswg.org/css-anchor-position-1/#swap-due-to-a-try-tactic</a>
[2]: <a href="https://github.com/w3c/csswg-drafts/issues/11828">https://github.com/w3c/csswg-drafts/issues/11828</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-013-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area-expected.txt:
* Source/WebCore/rendering/style/PositionArea.cpp:
(WebCore::operator&lt;&lt;):
    - Implemented operator&lt;&lt;(TextStream ts, PositionArea{Axis,Track,Self})
      operator&lt;&lt;(TextStream ts, PositionAreaSpan) will use these functions instead.
    - operator&lt;&lt; that takes references were changed to take values.

* Source/WebCore/rendering/style/PositionArea.h:
(WebCore::oppositePositionAreaAxis):
    - See below.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::flipPositionAreaByLogicalAxis):
    - Added function to flip position-area by a logical axis (block/inline).

(WebCore::Style::mirrorPositionAreaAcrossDiagonal):
    - Added function to flip position-area according to flip-start.

(WebCore::Style::BuilderConverter::convertPositionArea):
    - Added code to flip position-area according to the fallback try tactic,
      if specified.

(WebCore::Style::BuilderConverter::convertPositionTryFallbacks):
    - Tightened parsing a bit by rejecting the CSS value if duplicate
      tactics are found.

(WebCore::Style::positionAreaOppositeAxis): Deleted.
    - Renamed to oppositePositionAreaAxis, and moved to PositionArea.h

Canonical link: <a href="https://commits.webkit.org/293219@main">https://commits.webkit.org/293219@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/269ddfc12c144340cf06cf5f35fc2a6653f058ee

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98175 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8033 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103292 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100220 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18098 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26257 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74745 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31918 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101179 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13717 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88700 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55105 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13499 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6645 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48146 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83472 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6724 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105668 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25261 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18399 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83731 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25634 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84889 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83190 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21024 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27828 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5512 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18901 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25220 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30394 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25040 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28356 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26615 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2672,14 +2684,33 @@ inline std::optional<PositionArea> BuilderConverter::convertPositionArea(Builder
+            case WebCore::Style::PositionTryFallback::Tactic::FlipInline:
+                area = flipPositionAreaByLogicalAxis(LogicalBoxAxis::Inline, area, builderState.style().writingMode());
+                break;
+            case WebCore::Style::PositionTryFallback::Tactic::FlipStart:
```
You can probably drop `WebCore::Style` since you're already in that namespace?

tuankiet65 replied:
Yeah it got added by `clangd` suggestions 😃 I'll remove it

tuankiet65 replied:
Fixed.

anttijk commented on Source/WebCore/rendering/style/PositionArea.h:
```diff
@@ -47,6 +47,32 @@ enum class PositionAreaAxis : uint8_t {
+WTF::TextStream& operator<<(WTF::TextStream&, PositionAreaAxis);
+
+// Get the opposite axis of a given axis.
+static inline PositionAreaAxis flipPositionAreaAxis(PositionAreaAxis axis)
```
No need to inline this.

tuankiet65 replied:
It's a function declaration in a header file so the `inline` is needed.

anttijk commented on Source/WebCore/rendering/style/PositionTryFallback.h:
```diff
@@ -43,7 +43,7 @@ struct PositionTryFallback {
         FlipStart
     };
-    Vector<Tactic> tactics { };
+    Vector<Tactic, 3> tactics { };
```
Why this change in this patch? It is not really related.

tuankiet65 replied:
I've removed it.

anttijk commented on Source/WebCore/rendering/style/PositionArea.cpp:
```diff
@@ -164,40 +164,54 @@ ItemPosition PositionArea::defaultAlignmentForAxis(BoxAxis physicalAxis, Writing
     case PositionAreaAxis::Y:          ts << 'y'; break;
     case PositionAreaAxis::Block:      ts << "block"_s; break;
     case PositionAreaAxis::Inline:     ts << "inline"_s; break;
+    default:                           ASSERT_NOT_REACHED(); break;
```
Do we really need these? Lets not add random asserts to debug code.

tuankiet65 replied:
I've remove those.

tuankiet65 committed:
[css-anchor-position-1] Implement swapping due to a try-tactic for position-area
rdar://146814606
https://bugs.webkit.org/show_bug.cgi?id=289589

Reviewed by Antti Koivisto.

This commit implements swapping due to a try-tactic [1] for position-area.
This is done at style building time; BuilderConverter::convertPositionArea
will flip position-area as specified by the position-try fallback. This is
needed instead of flipping position-area at layout time, because
position-area should compute to the flipped value.

There are some remaining test failures in try-tactic-position-area.html;
these has to do with how position-area should be computed. The test expects
the computed value of position-area to use axis-ambiguous keywords, when the
specified value also uses those keywords (e.g: 'start end' computes to
'start end'). However, WebKit currently computes to their axis-unambiguous
equivalent (e.g: 'start end' computes to 'block-start block-end').
This issue is tracked in [2].

[1]: https://drafts.csswg.org/css-anchor-position-1/#swap-due-to-a-try-tactic
[2]: https://github.com/w3c/csswg-drafts/issues/11828

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-013-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-scroll-position-try-014-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-basic-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-change-fallbacks-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/last-successful-intermediate-ignored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/position-area-in-position-try-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/try-tactic-position-area-expected.txt:
* Source/WebCore/rendering/style/PositionArea.cpp:
(WebCore::operator<<):
    - Implemented operator<<(TextStream ts, PositionArea{Axis,Track,Self})
      operator<<(TextStream ts, PositionAreaSpan) will use these functions instead.
    - operator<< that takes references were changed to take values.

* Source/WebCore/rendering/style/PositionArea.h:
(WebCore::oppositePositionAreaAxis):
    - See below.

* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::flipPositionAreaByLogicalAxis):
    - Added function to flip position-area by a logical axis (block/inline).

(WebCore::Style::mirrorPositionAreaAcrossDiagonal):
    - Added function to flip position-area according to flip-start.

(WebCore::Style::BuilderConverter::convertPositionArea):
    - Added code to flip position-area according to the fallback try tactic,
      if specified.

(WebCore::Style::BuilderConverter::convertPositionTryFallbacks):
    - Tightened parsing a bit by rejecting the CSS value if duplicate
      tactics are found.

(WebCore::Style::positionAreaOppositeAxis): Deleted.
    - Renamed to oppositePositionAreaAxis, and moved to PositionArea.h

Canonical link: https://commits.webkit.org/293219@main
